pkg/kubelet/cm/devicemanager/manager.go

```diff
@@ -157,7 +157,7 @@ func newManagerImpl(socketPath string, topology []cadvisorapi.Node, topologyAffi
 		numaNodes:             numaNodes,
 		topologyAffinityStore: topologyAffinityStore,
 		devicesToReuse:        make(PodReusableDevices),
-		update:                make(chan resourceupdates.Update),
+		update:                make(chan resourceupdates.Update, 8),
```

Comment:
Is it ok that we use a length of 8? Because I have seen other chan's with a length of 20. Is there any inconsistency?

Reply:
because a node generally having 8 nvidia gpu or other device.

Reply:
what other examples do we have in code?

8 sounds like a small number here? What are downisdes of a bigger value here?

Reply:
@lengrongfu Would it make sense to send updates to the channel in a goroutine to avoid blocking `genericDeviceUpdateCallback` ?

Reply:
> what other examples do we have in code?

 I didn't find any device related chan, I found that pleg chan buffer is 1000.
https://github.com/kubernetes/kubernetes/blob/08382213e83b5e90cbfb9465d48ba67d51e7a7a0/pkg/kubelet/kubelet.go#L172

@googs1025 can you provide a buffer length is 20 code.

> 8 sounds like a small number here? What are downisdes of a bigger value here?

Larger values ​​don't seem to have a disadvantage, but values ​​within a reasonable range are probably better.

Reply:
> @lengrongfu Would it make sense to send updates to the channel in a goroutine to avoid blocking `genericDeviceUpdateCallback` ?

I think it makes sense. 
Because device-plugin performs device update through Send, if genericDeviceUpdateCallback is always blocked, device-plugin will time out.

Reply:
it would make sense. We can start with something simple. But we may need to have some sort of update queue deduplication logic so device plugin will not be able to DOS the kubelet and API server

Reply:
I would suggest this PR starts separate go routines. And if you can file an issue for the future on preventing DOS - we can follow up on this later

Reply:
Thanks, do you think there is anything else that needs to be modified in the current PR?

Reply:
the unit test that will ensure that there is no synchronization when channel is not being read would be helpful. Like call this 100 times and make sure all callls completed and then check that 100 updates can be read from the channel.

Reply:
done.

Reply:
sorry, I do not understand. Last state of this PR is not moving channel update into the go routine. Only set buffer to 8. But how the test with 100 elements works than? I meant let's do adding to channel in go routines. And test will validate that even with the small channel buffer, those calls will not be blocked

Reply:
got it, it like this

```go
update := make(chan resourceupdates.Update, 8)
go func() {
         for i := 0; i < 100; i++ {
		update <- resourceupdates.Update{}
	     }
	}()
for u := range update {
   	
}
```

Reply:
I am not sure now. 

So we need two properties of this:

1. Do not block the `genericDeviceUpdateCallback` so the device plugin's call will not fail with timeout
2. Protect kubelet from the DOS.

My new suggestion is to make a channel's buffer be like `100`. And then when we write to channel - drop the item if the channel is full (and log it as an error).

Reply:
Just to be clear, I was suggesting this change:
```
--- a/pkg/kubelet/cm/devicemanager/manager.go
+++ b/pkg/kubelet/cm/devicemanager/manager.go
@@ -309,9 +309,11 @@ func (m *ManagerImpl) genericDeviceUpdateCallback(resourceName string, devices [
 
        if utilfeature.DefaultFeatureGate.Enabled(features.ResourceHealthStatus) {
                if len(podsToUpdate) > 0 {
-                       m.update <- resourceupdates.Update{
-                               PodUIDs: podsToUpdate.UnsortedList(),
-                       }
+                       go func() {
+                               m.update <- resourceupdates.Update{
+                                       PodUIDs: podsToUpdate.UnsortedList(),
+                               }
+                       }()
                }
        }
```

And I still thinks it makes sense to do it to unblock genericDeviceUpdateCallback. As for the unit tests - I'm not sure I fully understood the idea of using goroutine there.

Reply:
In genericDeviceUpdateCallback use goroutine to not block device plugin's call, what do you think of this suggestion? @SergeyKanzhelev 

I think if we use goroutine, we need to consider DOS and avoid starting too many goroutines. 

> My new suggestion is to make a channel's buffer be like 100. And then when we write to channel - drop the item if the channel is full (and log it as an error).

I agree with this suggestion.

drop the item if the channel is full is also a way to protect kubelet from the DOS.


I was suggesting this change:
```
--- a/pkg/kubelet/cm/devicemanager/manager.go
+++ b/pkg/kubelet/cm/devicemanager/manager.go
@@ -309,9 +309,11 @@ func (m *ManagerImpl) genericDeviceUpdateCallback(resourceName string, devices [
 
        if utilfeature.DefaultFeatureGate.Enabled(features.ResourceHealthStatus) {
                if len(podsToUpdate) > 0 {
-                       m.update <- resourceupdates.Update{
-                               PodUIDs: podsToUpdate.UnsortedList(),
-                       }
+                       go func() {
+                              select {
+                                case m.update <- resourceupdates.Update{PodUIDs: podsToUpdate.UnsortedList()}:
+                                default:
+                                  fmt.Println("Channel is full")
+                             }
+                              
+                       }()
                }
        }
```

Reply:
yes, suggested code is what I envisioned

Reply:
Ok, modification completed.